Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrBridgeBuilder.kt

```diff
@@ -0,0 +1,273 @@
+            }
+            simpleStubContainer.properties.forEach {
+                it.accept(this, simpleStubContainer)
+            }
+            simpleStubContainer.typealiases.forEach {
+                it.accept(this, simpleStubContainer)
+            }
+            simpleStubContainer.simpleContainers.forEach {
+                it.accept(this, simpleStubContainer)
+            }
+        }
+    }
+
+    private fun isCValuesRef(type: StubType): Boolean {
+        if (type !is WrapperStubType) return false
+
+        return type.kotlinType is KotlinClassifierType && type.kotlinType.classifier == KotlinTypes.cValuesRef
+    }
+
+    private fun generateBridgeBody(function: FunctionStub) {
+        assert(function.origin is StubOrigin.Function) { "Can't create bridge for ${function.name}" }
+        val origin = function.origin as StubOrigin.Function
+        val bodyGenerator = KotlinCodeBuilder(scope = kotlinFile)
+        val bridgeArguments = mutableListOf<TypedKotlinValue>()
+        var isVararg = false
+        function.parameters.forEachIndexed { index, parameter ->
+            isVararg = isVararg or parameter.isVararg
+            val bridgeArgument = when {
+                parameter in builderResult.bridgeGenerationComponents.cStringParameters -> {
+                    bodyGenerator.pushMemScoped()
+                    "${parameter.name}?.cstr?.getPointer(memScope)"
+                }
+                parameter in builderResult.bridgeGenerationComponents.wCStringParameters -> {
+                    bodyGenerator.pushMemScoped()
+                    "${parameter.name}?.wcstr?.getPointer(memScope)"
+                }
+                isCValuesRef(parameter.type) -> {
+                    bodyGenerator.pushMemScoped()
+                    bodyGenerator.getNativePointer(parameter.name)
+                }
+                else -> {
+                    parameter.name
+                }
+            }
+            bridgeArguments += TypedKotlinValue(origin.function.parameters[index].type, bridgeArgument)
+        }
+        // TODO: Improve assertion message.
+        assert(!isVararg || context.platform != KotlinPlatform.NATIVE) {
+            "Function ${function.name} was processed incorrectly."
+        }
+        val result = mappingBridgeGenerator.kotlinToNative(
+                bodyGenerator,
+                function,
+                origin.function.returnType,
+                bridgeArguments,
+                independent = false
+        ) { nativeValues ->
+            "${origin.function.name}(${nativeValues.joinToString()})"
+        }
+        bodyGenerator.returnResult(result)
+        functionBridgeBodies[function] = bodyGenerator.build()
+    }
+
+    private fun generateProtocolGetter(protocolGetterName: String, protocol: ObjCProtocol) {
+        val builder = NativeCodeBuilder(simpleBridgeGenerator.topLevelNativeScope)
+        val nativeBacked = object : NativeBacked {}
```

Comment:
It would be great to have the same `TODO` here as before:
https://github.com/JetBrains/kotlin-native/pull/3110/files#diff-d16ba15be14e492e6988b703670530dcL515